In [14]:
import pandas as pd
import numpy as np
import re

pd.set_option("display.max_rows", 1000)

In [4]:
df = pd.read_csv("GSAF5.csv", encoding="ISO-8859-1")

After loading the dataset, I try to get a feeling for what's in there by checking the columns and the first few rows

In [9]:
df.head()

Case Number       Date  Year        Type    Country      Area  \
0  2016.09.18.c  18-Sep-16  2016  Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016  Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016  Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016  Unprovoked  AUSTRALIA  Victoria   
4    2016.09.15  16-Sep-16  2016  Unprovoked  AUSTRALIA  Victoria   

                           Location Activity             Name Sex   ...  \
0  New Smyrna Beach, Volusia County  Surfing             male    M  ...   
1  New Smyrna Beach, Volusia County  Surfing   Chucky Luciano    M  ...   
2  New Smyrna Beach, Volusia County  Surfing             male    M  ...   
3                  Thirteenth Beach  Surfing  Rory Angiolella    M  ...   
4                       Bells Beach  Surfing             male    M  ...   

    Species        Investigator or Source                        pdf  \
0        NaN  Orlando Sentinel, 9/19/2016       2016.09.18.c-NSB.pdf   
1        NaN  Orlando Sentinel, 9/19/2016   2016.09.18.b-Luciano.pdf   
2        NaN  Orlando Sentinel, 9/19/2016       2016.09.18.a-NSB.pdf   
3        NaN           The Age, 9/18/2016  2016.09.17-Angiolella.pdf   
4  2 m shark           The Age, 9/16/2016  2016.09.16-BellsBeach.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.c   
1  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.b   
2  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.a   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2016.09.17   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2016.09.16   

  Case Number.2 original order Unnamed: 22 Unnamed: 23  
0  2016.09.18.c           5993         NaN         NaN  
1  2016.09.18.b           5992         NaN         NaN  
2  2016.09.18.a           5991         NaN         NaN  
3    2016.09.17           5990         NaN         NaN  
4    2016.09.15           5989         NaN         NaN  

[5 rows x 24 columns]

In [8]:
df.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

To further examine the date columns ("Case Number", "Date", "Year"), I might print the unique values

In [17]:
df["Date"].value_counts().sort_index()

    10-Jan-2009    1
    15-Jun-1937    1
    16-Jan-1970    1
    22-Jul-2013    1
   21-Sep-1908     1
                  ..
Summer-2008        1
Winter 1942        1
Winter 1969        1
Woirld War II      1
World War II       2
Name: Date, Length: 5128, dtype: int64

We see that there are a bunch of weird values in the column

Test if we can use the pd.to_datetime() method on a bunch of rows to convert them to datetime.

In [22]:
pd.to_datetime(df.loc[0, "Date"], infer_datetime_format=True)

Timestamp('2016-09-18 00:00:00')

In [25]:
def are_you_inferable(date):
    if pd.to_datetime(date, infer_datetime_format=True):
        return True
    else:
        False

In [26]:
are_you_inferable("10-Jan-2009")

True

In [29]:
df["Date"].apply(are_you_inferable)

ValueError: ('Unknown string format:', 'Reported  14-Jul-2016')

This is obviously not working. We have to control for the error. If there only was a way to somehow capture it

In [30]:
def are_you_inferable2(date):
    try:
        pd.to_datetime(date, infer_datetime_format=True)
        return True
    except:
        return False

In [31]:
are_you_inferable2("Reported  14-Jul-2016")

False

In [32]:
df["Date"].apply(are_you_inferable2)

0        True
1        True
2        True
3        True
4        True
        ...  
5987    False
5988    False
5989    False
5990    False
5991    False
Name: Date, Length: 5992, dtype: bool

Now that we can ask every single row if it is inferable, we can easily quantify the two groups

In [33]:
df["Date"].apply(are_you_inferable2).value_counts()

True     5135
False     857
Name: Date, dtype: int64

Move the 857 rows that could not be infered into a separate df for further inspection

In [38]:
df_ninfer = df[df["Date"].apply(are_you_inferable2) == False].copy()
df_ninfer.head()

Case Number                   Date  Year        Type         Country  \
33   2016.07.14.4  Reported  14-Jul-2016  2016  Unprovoked         BAHAMAS   
34   2016.07.08.R   Reported 08-Jul-2016  2016  Unprovoked           SPAIN   
85   2016.03.03.R   Reported 03-Mar-2016  2016  Unprovoked       AUSTRALIA   
90   2016.02.10.R   Reported 10-Feb-2016  2016     Invalid  CAYMAN ISLANDS   
100  2016.01.11.R   Reported 11-Jan-2016  2016  Unprovoked       AUSTRALIA   

                Area                       Location            Activity  \
33               NaN                    Tiger Beach        Scuba Diving   
34    Canary Islands        Las Teresitas, Tenerife              Wading   
85   South Australia                    Wrights Bay             Fishing   
90      Grand Cayman              Stingray City Bar  Feeding stingrays?   
100       Queensland  Happy Valley Beach, Caloundra             Surfing   

                  Name Sex   ...              Species   \
33   Michael Dornellas    M  ...       Lemon shark, 9'   
34              female    F  ...           Angel shark   
85          Lee Taplin    M  ...         Bronze whaler   
90     Richard Branson    M  ...  No shark involvement   
100       Shane Hilder    M  ...       Wobbegong shark   

            Investigator or Source                                pdf  \
33              GrindTV, 7/14/2016        2016.07.14.R-TigerBeach.pdf   
34          La Opinion de Tenerife             2016.07.08.R-Spain.pdf   
85                9 News, 3/1/2016            2016.03.03.R-Taplin.pdf   
90                      R. Branson  2016.02.10.R-Branson-stingray.pdf   
100  ABC Sunshine Coast, 1/11/2016            2016.01.11.R-Hilder.pdf   

                                          href formula  \
33   http://sharkattackfile.net/spreadsheets/pdf_di...   
34   http://sharkattackfile.net/spreadsheets/pdf_di...   
85   http://sharkattackfile.net/spreadsheets/pdf_di...   
90   http://sharkattackfile.net/spreadsheets/pdf_di...   
100  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                  href Case Number.1  \
33   http://sharkattackfile.net/spreadsheets/pdf_di...  2016.07.14.R   
34   http://sharkattackfile.net/spreadsheets/pdf_di...  2016.07.08.R   
85   http://sharkattackfile.net/spreadsheets/pdf_di...  2016.03.03.R   
90   http://sharkattackfile.net/spreadsheets/pdf_di...  2016.02.10.R   
100  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.01.11.R   

    Case Number.2 original order Unnamed: 22 Unnamed: 23  
33   2016.07.14.4           5960         NaN         NaN  
34   2016.07.08.R           5959         NaN         NaN  
85   2016.03.03.R           5908         NaN         NaN  
90   2016.02.10.R           5903         NaN         NaN  
100  2016.01.11.R           5893         NaN         NaN  

[5 rows x 24 columns]

In [39]:
df_ninfer.shape

(857, 24)

In [40]:
df_ninfer["Date"].value_counts()

No date                                                             6
1970s                                                               5
No date, Before 1963                                                5
1960s                                                               4
Before 1958                                                         4
Before 1906                                                         4
Reported 10-Oct-1906                                                4
Before 1962                                                         3
May-Jun-1965                                                        3
1700s                                                               3
1941-1945                                                           3
Before 1911                                                         3
1940 - 1950                                                         3
Before 1921                                                         3
Jul-06              

In [41]:
df_ninfer[["Date", "Case Number"]]

Date         Case Number
33                                Reported  14-Jul-2016        2016.07.14.4
34                                 Reported 08-Jul-2016        2016.07.08.R
85                                 Reported 03-Mar-2016        2016.03.03.R
90                                 Reported 10-Feb-2016        2016.02.10.R
100                                Reported 11-Jan-2016        2016.01.11.R
151                                              Sep-15          2015.09.00
184                                Reported 25-Jun-2015        2015.06.25.R
247                                Reported 23-Dec-2014        2014.12.23.R
249                                Reported 03-Dec-2014        2014.12.03.R
254                                Reported 17-Nov-2014          2014.11.17
276                                Reported 12-Sep-2014        2014.09.13.R
282                                              Sep-14          2014.09.00
290                                Reported 25-Aug-2014          2014.08.25
303                                              Aug-14          2014.08.00
316                                Reported 27-Jun-2014        2014.06.27.R
320                                Reported 17-Jun-2014        2014.06.17.R
337                                Reported 10-May-2014        2014.05.10.R
343                                Reported 12-Apr-2014        2014.04.12.R
359                                Reported 17-Feb-2014          2014.02.17
419                                Reported 08-Aug-2013        2013.08.08.R
430                                Reported 17-Jul-2013        2013.07.17.R
445                                Reported 14-Jun-2013        2013.06.14.R
465                                Reported 02-Apr-2013        2013.04.02.R
469                                Reported 21-Mar-2013        2013.03.21.R
485                                Reported 21-Jan-2013        2013.01.21.R
496                                              Dec-12          2012.12.00
510                                Reported 11-Oct-2012        2012.10.11.R
548                                Reported 28-Jun-2012        2012.06.28.R
579                                              Apr-13          2012.04.00
597                                Reported 22-Jan-2012        2012.01.22.R
607                                Reported 26-Dec-2011        2011.12.26.R
621                                Reported 20-Nov-2011        2011.11.20.R
625                                Reported 28-Oct-2011        2011.10.28.R
662                                        16-Aug--2011        2011.08.16.b
665                                        11-Aug--2011          2011.08.11
688                                Reported 14-Jun-2011        2011.06.14.R
690                                Reported 06-Jun-2011        2011.06.06.R
703                                Reported 07-May-2011        2011.05.07.R
711                                Reported 29-Mar-2011        2011.03.29.R
717                                Reported 10-Mar-2010        2011.03.10.R
719                                Reported 28-Feb-2011        2011.02.28.R
725                                Reported 04-Feb-2011        2011.02.04.R
732                                Reported 12-Jan 2011        2011.01.12.R
738                                Reported 03-Dec-2010        2010.12.03.R
743                                Reported 27-Nov-2010        2010.11.27.R
748                                Reported 12-Nov-2010        2010.11.12.R
750                                Reported 28-Oct-2010        2010.10.28.R
766                                Reported 06-Sep-2010        2010.09.06.R
801                                              Jun-10          2010.06.00
813                                         190Feb-2010          2010.02.19
818                                Reported 06-Feb-2010        2010.02.06.R
855                                Reported 29-Oct-2009        2009.10.29.R
861                    

add some notes on what's happening with the case number column

In [46]:
date_pattern = r"(\d{4}\W\d{2}\W\d{2})"
df_ninfer["Case Number"].str.extract(date_pattern, expand=False)

33      2016.07.14
34      2016.07.08
85      2016.03.03
90      2016.02.10
100     2016.01.11
151     2015.09.00
184     2015.06.25
247     2014.12.23
249     2014.12.03
254     2014.11.17
276     2014.09.13
282     2014.09.00
290     2014.08.25
303     2014.08.00
316     2014.06.27
320     2014.06.17
337     2014.05.10
343     2014.04.12
359     2014.02.17
419     2013.08.08
430     2013.07.17
445     2013.06.14
465     2013.04.02
469     2013.03.21
485     2013.01.21
496     2012.12.00
510     2012.10.11
548     2012.06.28
579     2012.04.00
597     2012.01.22
607     2011.12.26
621     2011.11.20
625     2011.10.28
662     2011.08.16
665     2011.08.11
688     2011.06.14
690     2011.06.06
703     2011.05.07
711     2011.03.29
717     2011.03.10
719     2011.02.28
725     2011.02.04
732     2011.01.12
738     2010.12.03
743     2010.11.27
748     2010.11.12
750     2010.10.28
766     2010.09.06
801     2010.06.00
813     2010.02.19
818     2010.02.06
855     2009.10.29
861     2009

In [53]:
df_ninfer["Case Number"].str.extract(date_pattern, expand=False).apply(are_you_inferable2).value_counts()

True     640
False    217
Name: Case Number, dtype: int64

We're getting closer.
Create another df with the remaining rows

In [54]:
df_ninfer_ncase = df_ninfer[df_ninfer["Case Number"]
                            .str.extract(date_pattern, expand=False)
                            .apply(are_you_inferable2) == False].copy()

In [59]:
df_ninfer_ncase.sample(frac=1).head()

Case Number                    Date  Year        Type  Country  \
5769  1847.00.00.a                Ca. 1847  1847  Unprovoked      USA   
4868    1928.09.00                  Sep-28  1928  Unprovoked      USA   
2289  1993.00.00.c  Between May & Nov-1993  1993  Unprovoked  SOMALIA   
4266  1950.00.00.e             Summer 1950  1950  Unprovoked   GREECE   
3188  1970.00.00.c                   1970s  1970  Unprovoked     IRAQ   

                 Area                                Location  \
5769   South Carolina                                     NaN   
4868            Texas                          Corpus Christi   
2289  Banaadir Region                               Mogadishu   
4266              NaN                         Piraeus, Athens   
3188           Basrah  Shatt-al-Arab River near Abu al Khasib   

                       Activity                     Name Sex   ... Species   \
5769                    Boating                    adult    M  ...      NaN   
4868                    Fishing            W.R. Loesberg    M  ...      NaN   
2289                        NaN  several Somali children  NaN  ...      NaN   
4266                   Swimming                      NaN  NaN  ...      NaN   
3188  Washing clothes on stairs                   female    F  ...      NaN   

                                 Investigator or Source  \
5769                                    W.H. Greg, p.21   
4868                    Galveston Daily News, 12/7/1929   
2289                 Orlando Sentinel, 112/1/1993, p.A9   
4266  V.M. Coppleson (1958), p.259; L. Schultz & M. ...   
3188                       B.W. Coad & L.A.J. Al-Hassan   

                                     pdf  \
5769      1847.00.00.a-SouthCarolina.pdf   
4868             1928.09.00-Loesberg.pdf   
2289  1993.00.00.c - Somali children.pdf   
4266            1950.00.00.e-Piraeus.pdf   
3188       1970.00.00.c-Abu-a-Khasib.pdf   

                                           href formula  \
5769  http://sharkattackfile.net/spreadsheets/pdf_di...   
4868  http://sharkattackfile.net/spreadsheets/pdf_di...   
2289  http://sharkattackfile.net/spreadsheets/pdf_di...   
4266  http://sharkattackfile.net/spreadsheets/pdf_di...   
3188  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1  \
5769  http://sharkattackfile.net/spreadsheets/pdf_di...  1847.00.00.a   
4868  http://sharkattackfile.net/spreadsheets/pdf_di...    1928.09.00   
2289  http://sharkattackfile.net/spreadsheets/pdf_di...  1993.00.00.c   
4266  http://sharkattackfile.net/spreadsheets/pdf_di...  1950.00.00.e   
3188  http://sharkattackfile.net/spreadsheets/pdf_di...  1970.00.00.c   

     Case Number.2 original order Unnamed: 22 Unnamed: 23  
5769  1847.00.00.a            224         NaN         NaN  
4868    1928.09.00           1125         NaN         NaN  
2289  1993.00.00.c           3705         NaN         NaN  
4266  1950.00.00.e           1727         NaN         NaN  
3188  1970.00.00.c           2805         NaN         NaN  

[5 rows x 24 columns]

Replace "00" with "01" in the "Case Number" column

In [63]:
df_ninfer_ncase["Case Number"]

151         2015.09.00
282         2014.09.00
303         2014.08.00
496         2012.12.00
579         2012.04.00
801         2010.06.00
954         2009.01.00
964         2008.12.00
987         2008.09.00
1016        2008.07.00
1074      2008.00.00.b
1075      2008.00.00.a
1090        2007.11.00
1142        2007.07.00
1161        2007.05.00
1179        2007.02.00
1201      2006.10.00.b
1202      2006.10.00.a
1223      2006.08.00.b
1224      2006.08.00.a
1402        2004.11.00
1424        2004.08.00
1500        2003.11.00
1524        2003.09.00
1536      2003.07.00.c
1537      2003.07.00.b
1538      2003.07.00.a
1547        2003.06.00
1553        2003.05.00
1590      2002.11.00.a
1714        2001.07.00
1727        2001.05.00
1749        2001.03.00
1763        2000.12.00
1793      2000.09.00.b
1803        2000.08.00
1820        2000.07.00
1829        2000.06.00
1846        2000.03.00
2150        1995.07.00
2234      1994.00.00.b
2272      1993.05.00.b
2288      1993.00.00.d
2289      1

In [68]:
(df_ninfer_ncase["Case Number"]
 .str.replace("\.00", ".01")
 .str.extract(date_pattern, expand=False)
 .apply(are_you_inferable2)).value_counts()

True     204
False     13
Name: Case Number, dtype: int64

In [69]:
df_ninfer_ncase_nrepl = df_ninfer_ncase[df_ninfer_ncase["Case Number"]
                                        .str.replace("\.00", ".01")
                                        .str.extract(date_pattern, expand=False)
                                        .apply(are_you_inferable2) == False].copy()

In [72]:
df.drop(df_ninfer_ncase_nrepl.index)

Int64Index([5855, 5856, 5857, 5858, 5859, 5860, 5861, 5862, 5863, 5864, 5865,
            5866, 5867],
           dtype='int64')

In [71]:
pd.to_datetime("1642.01.01", infer_datetime_format=True)

OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 1642-01-01 00:00:00

Apply all the transformations. 

In [74]:
df["Age"].value_counts()

17                         148
18                         145
19                         138
20                         136
15                         135
16                         134
21                         115
22                         113
24                         103
25                         101
14                          97
13                          91
26                          80
23                          80
27                          78
28                          77
29                          75
30                          74
12                          70
35                          64
32                          64
10                          51
31                          50
40                          50
38                          47
34                          45
43                          43
36                          41
33                          40
37                          37
39                          37
42                          36
9       